src/corehost/inc/pal.h

```diff
@@ -96,7 +98,7 @@ namespace pal
 
     bool realpath(string_t& path);
     bool file_exists(const string_t& path);
-    std::vector<pal::string_t> readdir(const string_t& path);
+    void readdir(const string_t& path, std::vector<pal::string_t>* list);
```

Comment:
Pointers: NULL check necessary. Both side caller/callee contract that it is clearly a reference. No temporaries are allowed.
References: Implicit contract at the call site. Temporaries are allowed. Somehow it looks like a value but it is a pointer. Saves dereferencing.

@anurse, there are lots of links as why this is preferred. I've included a few below:

See Parameter Ordering and Reference Arguments sections which has some good reasoning: 
https://google.github.io/styleguide/cppguide.html#Function_Parameter_Ordering
https://google.github.io/styleguide/cppguide.html#Reference_Arguments

See rule 120 and 121 here: http://www.holub.com/goodies/rules.html

See this thread for additional info: http://stackoverflow.com/questions/334856/are-there-benefits-of-passing-by-pointer-over-passing-by-reference-in-c

Mostly, because this is also the style in all the teams I've worked here at MS as well.